crypto/ring-signature/src/ring_signature/mod.rs

```diff
@@ -17,24 +17,34 @@ mod error;
 mod generator_cache;
 mod key_image;
 mod mlsag;
-mod rct_bulletproofs;
 
 pub use curve_scalar::*;
 pub use error::Error;
 pub use generator_cache::*;
 pub use key_image::*;
 pub use mlsag::*;
-pub use rct_bulletproofs::*;
 
 /// The base point for blinding factors used with all amount commitments
 pub const B_BLINDING: RistrettoPoint = RISTRETTO_BASEPOINT_POINT;
 
-lazy_static! {
-    /// Generators (base points) for Bulletproofs.
-    /// The `party_capacity` is the maximum number of values in one proof. It should
-    /// be at least 2 * MAX_INPUTS + MAX_OUTPUTS, which allows for inputs, pseudo outputs, and outputs.
-    pub static ref BP_GENERATORS: BulletproofGens =
-        BulletproofGens::new(64, 64);
+/// This is a structure which contains a pair of orthogonal generators for
+/// Pedersen commitments.
+/// This tracks `bulletproofs::PedersenGens`, but we do not import it, to avoid
+/// creating a dependency on the `bulletproofs` crate.
```

Comment:
This was a pretty good overall move to generalize our transaction math and allow interchangeable implementations. I think people have been informally advocating we move forward with making our core transaction more general  (i.e. things like RistrettoPublic --> PublicKey trait, etc.) and being able to build more specialized implementations like this for partners.

I know this was done for a specific purpose. Do you think it's worth doing this for more of our internal transaction logic?

Reply:
> Do you think it's worth doing this for more of our internal transaction logic?

I guess I think, we should focus on solving MobileCoin's problems, and not other people's problems.

You can always make code more generic and more swappable -- you can dream up an infinite number of ways that any given piece of code can be more configurable and so on. It's a potentially infinite amount of work.

If the MobileCoin project has no immediate use of making something generic, meaning like, there are actually two possible implementations, then it's probably a waste of time -- we would be solving someone else's problems rather than the problems that we actually have.

This also feeds into YAGNI, another important software engineering principle.

There may be a reason that we want to generalize our transaction implementation, for example, one thing that was discussed was, maybe we can have a generic MLSAG implementation, so that we can share code between Monero's MLSAG and our implementation so that there's less code on a hardware wallet or something.

But lacking a specific, immediate need, I don't think making things more generic is better. I think it's easy for engineers to convince themselves to work on making things more generic and conceptified, but from a slightly outside perspective, it's easy to see that what they are doing doesn't actually add value and is basically a waste of time.

I say this having written a ton of generic code myself in a lot of projects, and thought a lot afterwards about what changes matter and what changes don't.

Sometimes though it is definitely better to make things generic to promote code re-use, particularly for like utilities and such. But in those cases you usually have a clear compelling need to solve the same problem repeatedly, and the code is not actually more complicated to write in a generic way.

Reply:
The other thing I would say is, ZK-STARKs are still on the road map, and that is going to be a completely different set of transaction math. If that happens in, say, 1 year, then the amount of time for "fully-generic transaction math" to have any positive impact on the project seems really small.

Reply:
But yeah, suffice it to say, the `PublicKey` trait seems like something that we shouldn't have done, because 
(1) we currently have no implementations of it except Ristretto and x25519, but for all the places where we used it it's not conceivable that we would substitute Ristretto for x25519 or vice-versa. So it's not generic in an interesting or useful way to the project. It's not actually creating swapability that is useful to us.
(2) We almost never actually use the abstract `key_exchange` function, we usually use direct Ristretto APIs where we use Ristretto because the keys crate traits don't expose the APIs we need, like the direct scalar multiplications used for subaddress derivation, for MLSAG implementation, for KeyImage definitions, etc. etc.
(3) The keys crate is extremely opinionated about what traits a PublicKey "should" implement and a lot of this is arbitrary and probably wrong. For instance it's required that all PublicKey are compatible with `serde`, but there is no reason for this. The first version also decided that every `PublicKey` should support `Into<Vec<u8>>` but this is not very good, there is no standard for how it is being serialized, like in what format. Is this DER? Something else? This `Into<Vec<u8>>` has now been moved into `ReprBytes` but it's still basically there because the AKE stuff requires that keys can do this.
(4) The keys crate currently bakes in a dependency on `alloc`, and so an allocator. This is hostile to hardware wallets, where we need to have as small a footprint as possible in the firmware. Real crypto crates like `curve25519-dalek` don't have a hard dependency on this, so this is accidental complexity, and basically a mess that people will have to clean up later, or avoid entirely by just not using the keys crate. The entire point of making a generic API like this would be to support portability efforts, like hardware wallets for exampke, so the fact that it didn't end up doing that is telling.

I think a lot of the stuff about `keys` traits are sort of half baked and need to be reworked. We created this when we were all just learning Rust and there's a lot of accidental complexity, assumptions that we will need something that never came to pass, and just like, not adequate review around this code, because it wasn't considered important by enough people.